### PR TITLE
Added support for External ScrollView in RecyclerListView

### PR DESCRIPTION
**Pull request by siddhantsoni**

### CODE REVIEW

File: src/core/scrollcomponent/web/ScrollViewer.js [hunk 1]
```diff
@@ -14,7 +14,9 @@ export default class ScrollViewer extends React.Component {
         this._setRelevantOffset = this._setRelevantOffset.bind(this);
         this._onWindowResize = this._onWindowResize.bind(this);
 
-        this.scrollEvent = {offsetX: 0, offsetY: 0};
+        this.scrollEvent = {};
+        this.scrollEvent.nativeEvent.contentOffset.x = 0;
+        this.scrollEvent.nativeEvent.contentOffset.y = 0;
```
Comment (siddhantsoni) [inline src/core/scrollcomponent/web/ScrollViewer.js]: Made the contract for scroll event to have parity with native scroll event.

File: src/core/scrollcomponent/web/ScrollViewer.js [hunk 2]
```diff
@@ -156,11 +158,11 @@ export default class ScrollViewer extends React.Component {
...
             if (this.props.horizontal) {
-                this.scrollEvent.offsetY = 0;
-                this.scrollEvent.offsetX = this.refs.mainDiv.scrollLeft;
+                this.scrollEvent.nativeEvent.contentOffset.y = 0;
+                this.scrollEvent.nativeEvent.contentOffset.x = this.refs.mainDiv.scrollLeft;
             } else {
-                this.scrollEvent.offsetX = 0;
-                this.scrollEvent.offsetY = this.refs.mainDiv.scrollTop;
+                this.scrollEvent.nativeEvent.contentOffset.x = 0;
+                this.scrollEvent.nativeEvent.contentOffset.y = this.refs.mainDiv.scrollTop;
```
Comment (siddhantsoni) [inline src/core/scrollcomponent/web/ScrollViewer.js]: Made changes in onScroll method according to new contract for web

File: src/core/scrollcomponent/web/ScrollViewer.js [hunk 3]
```diff
@@ -14,7 +14,9 @@ export default class ScrollViewer extends React.Component {
         this._setRelevantOffset = this._setRelevantOffset.bind(this);
         this._onWindowResize = this._onWindowResize.bind(this);
 
-        this.scrollEvent = {offsetX: 0, offsetY: 0};
+        this.scrollEvent = {};
+        this.scrollEvent.nativeEvent.contentOffset.x = 0;
```
Comment (naqvitalha) [inline src/core/scrollcomponent/web/ScrollViewer.js]: clear NPE, why do this at all?

File: src/core/scrollcomponent/web/ScrollViewer.js [hunk 4]
```diff
@@ -143,11 +145,11 @@ export default class ScrollViewer extends React.Component {
     _windowOnScroll() {
         if (this.props.onScroll) {
             if (this.props.horizontal) {
-                this.scrollEvent.offsetY = 0;
-                this.scrollEvent.offsetX = window.scrollX - this.props.distanceFromWindow;
+                this.scrollEvent.nativeEvent.contentOffset.y = 0;
```
Comment (naqvitalha) [inline src/core/scrollcomponent/web/ScrollViewer.js]: NPE

File: src/core/RecyclerListView.js [hunk 5]
```diff
@@ -45,10 +46,11 @@ if (process.env.RLV_ENV && process.env.RLV_ENV === 'browser') {
     platform = "web";
     ScrollComponent = require("./scrollcomponent/web/ScrollComponent").default;
     ViewRenderer = require("./viewrenderer/web/ViewRenderer").default;
+    ScrollViewer = require("./scrollcomponent/web/ScrollViewer").default;
```
Comment (naqvitalha) [inline src/core/RecyclerListView.js]: Not what I had intended, I mean you should do it inside the ScrollComponent

File: src/core/scrollcomponent/web/ScrollViewer.js (hunk 4 above)
Comment (naqvitalha) [inline src/core/scrollcomponent/web/ScrollViewer.js]: You'll need to change how it is being consumed too

File: src/core/scrollcomponent/web/ScrollViewer.js (hunk 3 above)
Comment (siddhantsoni) [reply src/core/scrollcomponent/web/ScrollViewer.js]: Removed!

File: src/core/scrollcomponent/web/ScrollComponent.js [hunk 6]
```diff
@@ -31,6 +31,7 @@ class ScrollComponent extends React.Component {
     }
 
     render() {
+        let ScrollViewer = this.props.externalScrollViewer;
```
Comment (naqvitalha) [inline src/core/scrollcomponent/web/ScrollComponent.js]: const

File: src/core/scrollcomponent/web/ScrollViewer.js [hunk 7]
```diff
@@ -14,7 +14,7 @@ export default class ScrollViewer extends React.Component {
         this._setRelevantOffset = this._setRelevantOffset.bind(this);
         this._onWindowResize = this._onWindowResize.bind(this);
 
-        this.scrollEvent = {offsetX: 0, offsetY: 0};
+        this.scrollEvent = {};
```
Comment (naqvitalha) [inline src/core/scrollcomponent/web/ScrollViewer.js]: define the whole object or else it won't work

Comment (naqvitalha) [reply src/core/scrollcomponent/web/ScrollViewer.js]: {nativeEvent:{
                        contentOffset: {x:0, y:0}     
}
}

File: src/core/RecyclerListView.js [hunk 8]
```diff
@@ -467,6 +466,10 @@ RecyclerListView.propTypes = {
     //On scroll callback onScroll(rawEvent, offsetX, offsetY), note you get offsets no need to read scrollTop/scrollLeft
     onScroll: PropTypes.func,
 
+    //Provide your own ScrollView Component. The contract for the scroll event should match the native scroll event contract, i.e.
+    // event = { nativeEvent: { contentOffset: { x: offset, y: offset } } }
+    externalScrollViewer: PropTypes.func,
```
Comment (naqvitalha) [inline src/core/RecyclerListView.js]: call it externalScrollView and also publish the full contract. I think we need on sizeChanged too and onScroll event

File: src/core/scrollcomponent/reactnative/ScrollComponent.js [hunk 9]
```diff
@@ -39,8 +39,9 @@ class ScrollComponent extends React.Component {
     }
 
     render() {
+        let ScrollViewer = this.props.externalScrollViewer;
```
Comment (naqvitalha) [inline src/core/scrollcomponent/reactnative/ScrollComponent.js]: const

Comment (naqvitalha) [reply src/core/scrollcomponent/reactnative/ScrollComponent.js]: Call it ScrollView only

File: src/core/scrollcomponent/web/ScrollViewer.js [hunk 10]
```diff
@@ -143,11 +150,15 @@ export default class ScrollViewer extends React.Component {
     _windowOnScroll() {
         if (this.props.onScroll) {
             if (this.props.horizontal) {
-                this.scrollEvent.offsetY = 0;
-                this.scrollEvent.offsetX = window.scrollX - this.props.distanceFromWindow;
+                if(this.scrollEvent && this.scrollEvent.nativeEvent && this.scrollEvent.nativeEvent.contentOffset) {
```
Comment (naqvitalha) [inline src/core/scrollcomponent/web/ScrollViewer.js]: No need of null checks here, the reason NPE would've happened here was that the complete object was not declared above

**Pull request merged**